### PR TITLE
#160577478 Build Endpoint For Social Authentication

### PR DESCRIPTION
**Pull request by michael-basweti**
#### What does this PR do?
          Authenticate user using social media accounts i.e. Twitter, Facebook, Google
#### Description of Task to be completed?
         -Have social authentication for endpoint working
               /api/social_auth/
#### How should this be manually tested?
         -After cloning the repo, cd into it and run python mange.py runserver
         -in your postman put the following data depending the provider you are using:
                      {
    "provider": "facebook",
    "access_token": "EAAe6UMnIZBRoBADj5XxZBAIKILAs3SBZCZAR60OJUhOr9JFSmqBLDxs5t8FAc9m4XYsPOwF7BwIuG47FoRO04FOpwE9ZAL3g7sA0ZAxZA2DegMQOrs9wFzP0BttnJjfES6gt8qA08g1s9BuYLKR3OpempTleeg1OjyWB2hGYE8YQOTXq93cvVqsh1ibm6C2a1lplRFCugfsxvx4MKudRS6DMUuUeO3yemBZB7PTC6rxYuYtvfzi5QJtpm"
    
}
#### Any background context you want to provide?
       -when using twitter authentication,apart from "access_token" you also need "access_token_secret" because it uses oauth1
#### What are the relevant pivotal tracker stories?
           #160577474
           #160577475

### CODE REVIEW

**Review by IncredibleBlackMan — COMMENTED** [review]
@michael-basweti I have reviewed the work, but I do not understand the change in the database engine?

**Comment by gloriaodipo** [comment]
Hey @michael-basweti please include docstrings in your tests

**Review by codingedward — COMMENTED** [review]
Hi Micheal,
Have you rebased this atop the updated develop branch?  I see you have implemented `JWTAuthentication` class which has already been merged and this will cause a merge conflict.

As this PR relates to social authentication only, I hope we will keep the merged version when resolving the conflict.

File: authors/apps/authentication/views.py [hunk 1]
```diff
@@ -73,3 +78,68 @@ def update(self, request, *args, **kwargs):
...
+        backend = load_backend(strategy=strategy, name=provider, redirect_uri=None)
+
+
+        if isinstance(backend, BaseOAuth1):
+           #cater for services that use OAuth1, an example is twitter
+            token = {
+                'oauth_token': serializer.data['access_token'],
+                'oauth_token_secret': request.data['access_token_secret'],
+            }
+            print(token)
```
Comment (codingedward) [inline authors/apps/authentication/views.py]: Let's get rid of the print

File: authors/apps/authentication/views.py [hunk 2]
```diff
@@ -73,3 +78,68 @@ def update(self, request, *args, **kwargs):
...
+        elif isinstance(backend, BaseOAuth2):
+            #we just need to pass access_token for OAuth2
+            token = serializer.data['access_token']
+            print(token)
```
Comment (codingedward) [inline authors/apps/authentication/views.py]: This one too.

File: authors/apps/authentication/views.py (hunk 1 above)
Comment (michael-basweti) [reply authors/apps/authentication/views.py]: thanks @codingedward, I just removed them

File: authors/apps/authentication/views.py (hunk 2 above)
Comment (michael-basweti) [reply authors/apps/authentication/views.py]: I will remove this one too @codingedward

**Review by koechkevin — COMMENTED** [review]
I have gone through your implementation, I would only like to know if you have rebased this to develop.

**Comment by FawazFarid** [comment]
@michael-basweti did you rebase against develop? some of these changes look like they were manually added.

**Comment by michael-basweti** [comment]
@FawazFarid I have been rebasing  to develop everytime a PR is merged to develop

File: authors/settings.py [hunk 3]
```diff
@@ -142,14 +150,60 @@
...
+
+
+SOCIAL_AUTH_GOOGLE_OAUTH2_KEY = '6613413486293-mkb39ej2rlv4b7m0t6o1pian5dr8nffd.apps.googleusercontent.com'
+SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET = 'MY69flrLPH5fzUCJY2rGbzus'
+
+
+
 # Activate django-heroku for Heroku
-django_heroku.settings(locals())
+# django_heroku.settings(locals())
```
Comment (FawazFarid) [inline authors/settings.py]: lets remove commented out code, here

**Review by FawazFarid — CHANGES_REQUESTED** [review]

File: authors/settings.py [hunk 4]
```diff
@@ -142,14 +150,56 @@
...
     ),
 }
 
+AUTHENTICATION_BACKENDS = (
+    'social_core.backends.open_id.OpenIdAuth',
+    'social_core.backends.google.GoogleOpenId',
+    'social_core.backends.google.GoogleOAuth2',
+    'social_core.backends.google.GoogleOAuth',
+    'social_core.backends.twitter.TwitterOAuth',
+    #'social_core.backends.yahoo.YahooOpenId',
```
Comment (FawazFarid) [inline authors/settings.py]: We can remove this since we don't need it

File: authors/settings.py [hunk 5]
```diff
@@ -142,14 +150,56 @@
...
+    'social_core.pipeline.social_auth.social_user',
+    'social_core.pipeline.user.get_username',
+    'social_core.pipeline.social_auth.associate_by_email',
+    'social_core.pipeline.user.create_user',
+    'social_core.pipeline.social_auth.associate_user',
+    'social_core.pipeline.social_auth.load_extra_data',
+    'social_core.pipeline.user.user_details',
+)
+SOCIAL_AUTH_FACEBOOK_KEY = '2175180982515994'
+SOCIAL_AUTH_FACEBOOK_SECRET = '4a8f6c5eb2aed7056edfe7d4323f28bf'
```
Comment (FawazFarid) [inline authors/settings.py]: It is not good practice to commit (expose) private keys, passwords or other sensitive details to GitHub. You should instead, store them as environment variables.

File: authors/settings.py [hunk 6]
```diff
@@ -142,14 +150,56 @@
...
+
+SOCIAL_AUTH_TWITTER_KEY = 'Z99QEtxVAZNYu5qcPRiRlqxBj'
+SOCIAL_AUTH_TWITTER_SECRET = 'lhaAA8wWcWpCmhtEPpGlWJgNyP9r2fRAsVNxFsFSa35VNQl2Fx'
+
+SOCIAL_AUTH_FACEBOOK_SCOPE =['email']
+SOCIAL_AUTH_FACEBOOK_PROFILE_EXTRA_PARAMS={'fields':'id,email,name'}
```
Comment (FawazFarid) [inline authors/settings.py:186]: ditto

File: authors/settings.py [hunk 7]
```diff
@@ -142,14 +150,56 @@
...
+
+
+
+SOCIAL_AUTH_GOOGLE_OAUTH2_KEY = '6613413486293-mkb39ej2rlv4b7m0t6o1pian5dr8nffd.apps.googleusercontent.com'
+SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET = 'MY69flrLPH5fzUCJY2rGbzus'
+
```
Comment (FawazFarid) [inline authors/settings.py]: ditto

**Review by FawazFarid — CHANGES_REQUESTED** [review]

**Comment by FawazFarid** [comment]
Hey can you fix the merge conflicts?

File: Pipfile [hunk 8]
```diff
@@ -14,15 +14,21 @@ gunicorn = "==19.9.0"
...
-"psycopg2" = "==2.7.5"
-"psycopg2-binary" = "==2.7.5"
+psycopg2 = "==2.7.5"
+psycopg2-binary = "==2.7.5"
 pylint = "==2.1.1"
 six = "==1.11.0"
 whitenoise = "==4.1"
 wrapt = "==1.10.11"
 pytz = "==2018.5"
-Django = "==2.1.1"
```
Comment (FawazFarid) [inline Pipfile]: why is this removed?

File: authors/apps/profiles/test_user_profile.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        { 
+            'user' : { 
+                'username': 'iano', 
+                'bio': ''
+            }
+        }, format='json', HTTP_AUTHORIZATION=token)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertIn(b'This field may not be blank', response.content)
+
+    def test_update_profile_spaces_only_in__bio(self):
```
Comment (FawazFarid) [inline authors/apps/profiles/test_user_profile.py]: there are two underscores between `in` and `bio` can you remove one?

File: authors/apps/profiles/test_user_profile.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        token = self.registration()
+        self.client.post(self.url_register, self.user, format='json')
+        response = self.client.put(self.url_profile,
```
Comment (FawazFarid) [inline authors/apps/profiles/test_user_profile.py:120]: the indentation here is a bit off, please make it more readable

File: authors/settings.py [hunk 11]
```diff
@@ -45,7 +45,10 @@
     'authors.apps.authentication',
     'authors.apps.core',
     'authors.apps.profiles',
+
+
```
Comment (FawazFarid) [inline authors/settings.py]: these spaces seem unnecessary. I would get rid of them

File: authors/settings.py [hunk 12]
```diff
@@ -142,14 +149,49 @@
...
+SOCIAL_AUTH_FACEBOOK_SECRET = os.getenv("FB_SECRET_AUTH")
+
+SOCIAL_AUTH_TWITTER_KEY = os.getenv("TWITTER_KEY")
+SOCIAL_AUTH_TWITTER_SECRET = os.getenv("TWITTER_SECRET")
+
+SOCIAL_AUTH_FACEBOOK_SCOPE =['email']
+SOCIAL_AUTH_FACEBOOK_PROFILE_EXTRA_PARAMS={'fields':'id,email,name'}
+
+
+
```
Comment (FawazFarid) [inline authors/settings.py]: there are three empty lines here, Can you reduce them to one?

File: authors/urls.py [hunk 13]
```diff
@@ -22,6 +22,7 @@
 urlpatterns = [
     path('admin/', admin.site.urls),
     path('api/', include(('authors.apps.authentication.urls', 'authentication'), namespace='authentication')),
-    path(r'', schema_view)
+    path(r'', schema_view),
+    path('api/', include(('authors.apps.profiles.urls', 'profiles'), namespace='profiles')),
```
Comment (FawazFarid) [inline authors/urls.py:26]: This is good, but It would be better if the line `path('api/', include(('authors.apps.profiles.urls', 'profiles'), namespace='profiles')),` came after `path('api/', include(('authors.apps.authentication.urls', 'authentication'), namespace='authentication'))`, grouping the api endpoints together looks better and easy to find what you looking for.

I'd say put `path(r'', schema_view),` before the the admin endpoint, and the api endpoints after admin endpoint

**Review by FawazFarid — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 14]
```diff
@@ -104,3 +104,4 @@ db.sqlite3
  # vim python config
 .ycm_extra_conf.py
 
+.vscode/settings.json
```
Comment (FawazFarid) [inline .gitignore]: also this doesnt seem to ignore the .vscode folder

File: authors/apps/profiles/test_user_profile.py (hunk 10 above)
Comment (michael-basweti) [reply authors/apps/profiles/test_user_profile.py:120]: @FawazFarid I  thought indentation is supposed to be 4 spaces and they are 4 according to how Ian wrote the tests, I would appreciate if you would clarify on this.

File: authors/apps/profiles/test_user_profile.py (hunk 9 above)
Comment (michael-basweti) [reply authors/apps/profiles/test_user_profile.py]: resolved

File: Pipfile (hunk 8 above)
Comment (michael-basweti) [reply Pipfile]: resolved

File: authors/settings.py (hunk 12 above)
Comment (michael-basweti) [reply authors/settings.py]: resolved

File: authors/settings.py (hunk 11 above)
Comment (michael-basweti) [reply authors/settings.py]: resolved

File: .gitignore (hunk 14 above)
Comment (michael-basweti) [reply .gitignore]: actually the .vscode is in develop, it was merged to develop, I don't have it in my branch

File: Pipfile (hunk 8 above)
Comment (FawazFarid) [reply Pipfile]: can you set it to `2.1.2` instead? `2.1.1` will give us the Github warning

**Review by FawazFarid — APPROVED** [review]
